CORDA-3601 Record a flow's finish time

lankydan opened the pull request:
Record a flow's finish time by updating its metadata record. It is set
in `updateCheckpoint` by checking the status of the checkpoint. If it is
`COMPLETED` it will set the `finishInstant` on the metadata object and
 update it.

lankydan committed:
CORDA-3601 Record a flow's finish time

Record a flow's finish time by updating its metadata record. It is set
in `updateCheckpoint` by checking the status of the checkpoint. If it is
 `COMPLETED` it will set the `finishInstant` on the metadata object and
 update it.

williamvigorr3 reviewed (dismissed):
LGTM!

kyriathar commented on node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt:
```diff
@@ -355,6 +356,24 @@ class FlowFrameworkTests {
+            assertNull(metadata.finishInstant)
+        }
+        terminationSignal.release()
+        mockNet.waitQuiescent()
```
Maybe we could get rid of this `waitQuiescent()` if we `getOrThrow` instead on the returned fiber's future.

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -380,13 +380,20 @@ class DBCheckpointStorage(private val checkpointPerformanceRecorder: CheckpointP
 
+        val metadata = entity.flowMetadata.apply {
+            if (checkpoint.status == FlowStatus.COMPLETED && finishInstant == null) {
+                finishInstant = now
```
Interesting that we initialise `now` to `Instant.now()`.  `DBCheckpointStorage` should look at the `clock` on `ServiceHub`.  Can we pass that into the constructor and use `clock.instant()` (I think!) please?

rick-r3 replied:
I think we  should update for all "finish types", so COMPLETED/FAILED/KILLED

lankydan replied:
I have some changes locally to correct that (the clock stuff). Going to do it in a separate PR, was waiting for this to go in.

lankydan committed:
CORDA-3601 Record flow finish time for all finished statuses

Update the flow finish time for the following statuses:

- COMPLETED
- KILLED
- FAILED

lankydan committed:
CORDA-3601 Use platform clock in `DBCheckpointStorage`
